Add RzIL for floating-point x86 instructions

DMaroo opened the pull request:
<!-- Filling this template is mandatory -->

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

<!-- Explain the **details** for making this change. Is a new feature implemented? What existing problem does the pull request solve? How does the pull request solve these issues? Please provide enough information so that others can review your pull request. -->

Add RzIL implementation for x86 floating point instructions. This is a WIP currently.

~All the unrelated commits will go away once #3864 gets merged.~ Done.

**Test plan**

<!-- What steps should the reviewer take to test your pull request? Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots/videos. This is your time to re-check that everything works and that you covered all the edge cases -->

Added tests in `db/asm/x86_64`. Pretty much all the instructions are covered. May add more tests depending on the code coverage.

Tracetesting has been put on hold for now (see [this comment](https://github.com/rizinorg/rizin/pull/3865#issuecomment-1878265146)), but will add the pending tracetesting in the tracker issue.

**Closing issues**

<!-- put "closes #XXXX" in your comment to auto-close the issue that your PR fixes (if any). -->

Closes #3861.

DMaroo committed:
Create `il_fp_ops.inc`, for floating-point instructions' IL ops

DMaroo committed:
Add IL implementation for `FABS`

    * Also add `x86_il_{get,set}_st_reg` helper functions

DMaroo committed:
Add IL implementation for `FNINIT` and `FLDCW`

DMaroo committed:
Add IL implementations for `FNSTCW` and `FNSTSW`

DMaroo committed:
Add IL implementation for `FNCLEX`

DMaroo committed:
Add ST push and pop functions

DMaroo committed:
Fix clang formatting

DMaroo committed:
Add fucntions for updating FPSW TOP pointer

DMaroo committed:
Add `FLD` instruction lifitng, FPSW flag support, FP operand support

DMaroo committed:
Remove `x86_bool_to_bv` and use the standard `BOOL_TO_BV`

DMaroo committed:
Add implementation for `FST{P}`, and rounding mode support

DMaroo committed:
Add `FLD` variants for constants

DMaroo committed:
Fix formatting and build

DMaroo committed:
Add IL lifting for `FXCH`

DMaroo committed:
Add width checks to avoid redundant rounding

DMaroo committed:
Implement IL lifting for `FILD`, `FIST{P}`

    * Add helper functions to inter-convert floating point values and
      integers using the rounding mode in the control word

DMaroo committed:
FPU stack regs are 80-bit, not 64-bit

DMaroo committed:
Add RzIL lifting for `FBLD`

XVilka reviewed (changes requested):
I also recommend taking some binary and targeting FP instructions located in it. Either one already in rizin-testbins, or take/compile a new one. Then, it can be tested as early as a subset of the instructions used is implemented.

DMaroo committed:
Add RzIL implementation for `FBSTP`

    * An invalid-arithmetic-operand (#IA) exception may occurr if the
      value being converted to the BCD integer cannot fit in an 18-bit
      BCD integer. For the IL's purpose, we just jump to the "int" label
      (stands for interrupt) using `GOTO`.

DMaroo committed:
Remove all the `EMPTY()` ops after `GOTO()` ops

    * On second thought, using `EMPTY()` ops after `GOTO()` ops is not
      the best idea to indicate end of analysis. Maybe the hooks can
      signal that analysis needs to be restarted.
    * Also, clean up the comments of a bunch of unimplemented ops

DMaroo commented:
I'm planning to write a testing 2D matrix multiplication code for floating point numbers to test the IL. I am not done with the arithmetic FP operations, so I need to lift them first before adding such a test. Maybe some programs from here can also be used: https://benchmarksgame-team.pages.debian.net/benchmarksgame/index.html.

DMaroo committed:
Moved some code around

DMaroo committed:
Rename "rmode" local variable to "_rmode"

DMaroo committed:
Add some TODOs for deffered work

DMaroo committed:
Fix failing asm tests

    * Caused due to the removal of the [empty] opcode

DMaroo committed:
Implement IL lifting for `FADD` (and some other minor refactoring)

    * Implement the IL lifting for `FADD`: This is not entirely correct
      as of now, because we don't handle setting C1 based on rounding
      up/down, but I think it should be fine for now. Need to tackle
      setting the control word bits later
    * Start using `RzFloatFormat` to describe the float's width instead
      of using an unsigned integer. Provide and expose conversion
      functions between the two

DMaroo committed:
Use a global to denote when RMode needs to be init, implement FIADD

    * Other minor refactor which fixes the CI (hopefully)

DMaroo committed:
Implement `FMUL` and `FIMUL`

    * Extract the common code into a C macro

DMaroo committed:
Implement RzIL for `FSUB` and `FISUB`

    * Also add `x86_il_fsub_with_rmode` helper

DMaroo committed:
Implement RzIL for `FSUBR` and `FISUBR`

    * Reuse the implementation of `x86_il_fsub_with_rmode` and just pass
      the arguments in the opposite order

DMaroo committed:
Add RzIL implementation for `FDIV`, `FIDIV`, `FDIVR`, `FIDIVR`

DMaroo committed:
Add pop versions of floating point arithmetic instuctions

    * Add a new macro `X86_ARITHMETIC_POP_IL` macro for the same
    * Fix a typo
    * Add a note for using `X86_INS_PFADD` for FADDP instruction on
      Capstone version newer than 4

DMaroo committed:
Add RzIL implementations for `FCOM`, `FCOMP` and `FCOMPP`

DMaroo committed:
Add RzIL lifting for `FUCOM` and `FCOMI` families of instructions

DMaroo committed:
Implement RzIL lifting for `FCHS` and `FTST`

DMaroo committed:
Add RzIL implementation for `FRNDINT` and `FSQRT`

DMaroo committed:
Add RzIL implementation for `FNOP` and `FISTTP`

DMaroo commented:
50+ x86 floating point instructions have been lifted to RzIL! This is a significant majority of the x86 floating point instructions (discount SSE, MMX, AVX and other such ISA extensions). Most of the remaining instructions are either too complicated and rare to bother implementing or use transcendental operations unsupported by RzIL or are _lower-level_ instructions which would not have any equivalent reasonable representation in RzIL. I believe the lifted instructions are more than enough to allow for a very satisfactory usage experience.

I need to now add asm tests for all these instructions and then perform tracetesting. I am not adding any more instruction liftings until the testing of the existing liftings is over.

DMaroo committed:
Remove global variable `use_rmode` and use pass around a context instead

    * Define a new `X86ILContext` struct to pass around information
      whether we need to intitialize the rouding mode or not
    * Remove the global variable which was previously responsible for
      doing so

DMaroo committed:
Add RzIL for `FICOM` and `FICOMP`, add register bindings and fix bugs

DMaroo committed:
Make `rz-test` more robust against IL outputs with newlines in them

DMaroo committed:
Add RzIL tests in db/asm for x86 FPU instructions

DMaroo committed:
Fix the bug when using `FADDP` with Capstone version > 4

DMaroo commented:
ASM tests pass with the existing x86 FPU instructions in `db/asm`. Will probably add some more instructions in the test file before moving forward with tracetesting.

DMaroo committed:
Fix remaining db tests

DMaroo committed:
Remove tests for `FSTSW` or `FSTCW` instructions

    * We already test `WAIT` (in `db/asm/x86_32`) and `FSTSW` (or
      `FSTCW`) (in `db/asm/x86_64`)
    * Remove the newline replacement code in `rz-test` as well, since we
      wouldn't need that functionality anymore

DMaroo committed:
Add asm tests for math constant push instructions

DMaroo commented:
For tracetesting, I am waiting on BinaryAnalysisPlatform/qemu#24 to come through. Until then, I believe this PR is mature enough to be merged. I'll open a new PR for the tracetesting fixes once we shift to using QEMU TCG plugins (because rebasing the current changes on top of QEMU is a painful and futile effort).

XVilka reviewed (changes requested):
Requires "emulateme"-style test

Rot127 commented on librz/analysis/arch/x86/common.c:
```diff
@@ -1070,4 +1060,468 @@ RzILOpEffect *x86_il_set_flags(RZ_OWN RzILOpPure *val, unsigned int size) {
+ *
+ * \param reg
+ */
+bool x86_il_is_st_reg(X86Reg reg) {
```
If this functions and the ones below are not static they should be `RZ_IPI`, right?

Rot127 replied:
Also, please add `RZ_OWN`, `RZ_BORROW` and `const` where applicable

DMaroo replied:
I'll add the annotations, thank you.

Rot127 commented on librz/analysis/arch/x86/common.c:
```diff
@@ -1070,4 +1060,468 @@ RzILOpEffect *x86_il_set_flags(RZ_OWN RzILOpPure *val, unsigned int size) {
+ *
+ * \param int_val Desirable that it is a small expression since it will be duped
+ * \param format Output float format
+ * \param ctx
```
Missing description of `ctx`. Also for other functions.

Rot127 replied:
Especially relevant since this function does "something with the context".

Rot127 commented on librz/analysis/arch/x86/common.c:
```diff
@@ -1070,4 +1060,468 @@ RzILOpEffect *x86_il_set_flags(RZ_OWN RzILOpPure *val, unsigned int size) {
+ * \return RzILOpFloat*
+ */
+RzILOpFloat *x86_il_floating_from_int_ctx(RzILOpBitVector *int_val, RzFloatFormat format, X86ILContext *ctx) {
+	ctx->use_rmode = true;
```
Please add `rz_return_val_if_fail()` to all those helpers. Since the pointer are derefenced without being checked before.

DMaroo replied:
I'll add the annotations and early returns as well.

Rot127 commented on librz/analysis/arch/x86/common.c:
```diff
@@ -1070,4 +1060,468 @@ RzILOpEffect *x86_il_set_flags(RZ_OWN RzILOpPure *val, unsigned int size) {
+ * \param ctx
+ * \brief Subtract \p y from \p x (reverse of \ref x86_il_fsub_with_rmode)
+ */
+RzILOpFloat *x86_il_fsubr_with_rmode_ctx(RzILOpFloat *x, RzILOpFloat *y, X86ILContext *ctx) {
```
Order better change order to have  `\brief` first

Rot127 commented on librz/analysis/arch/x86/common.c:
```diff
@@ -1070,4 +1060,468 @@ RzILOpEffect *x86_il_set_flags(RZ_OWN RzILOpPure *val, unsigned int size) {
+
+/**
+ * \brief Set the value of FPU status word.
+ * See \ref x86_il_get_fpu_stack_top() for tehe structure of FPU status word and
```
```suggestion
 * See \ref x86_il_get_fpu_stack_top() for the structure of FPU status word and
```

Rot127 commented on librz/analysis/arch/x86/common.c:
```diff
@@ -1070,4 +1060,468 @@ RzILOpEffect *x86_il_set_flags(RZ_OWN RzILOpPure *val, unsigned int size) {
+	RzILOpPure *shifted_top = SHIFTL0(UNSIGNED(16, top), UN(8, 11));
+	/* 0x3800 only has the 12, 13 & 14 bits set, so we take its negation for the
+	 * mask. */
+	RzILOpPure *mask = UN(16, ~(0x3800));
```
If feasible replace the magic number (`0x3800`) of the mask with a descriptive macro.

DMaroo replied:
I don't think using a macro would help. We are only using this magic number at one place. Maybe writing out the computation (`((1 << 3) - 1) << 11`) may help, but I think that just makes it more complicated.

Rot127 commented on librz/analysis/arch/x86/il_fp_ops.inc:
```diff
@@ -0,0 +1,697 @@
+	return SEQ2(x86_il_set_st_reg(X86_REG_ST0, abs_value, RZ_FLOAT_IEEE754_BIN_80), x86_il_set_fpu_flag(X86_FPU_C1, IL_FALSE));
+}
+
+#define FLOATING_ARITHMETIC_IL(op) \
```
More a decision of taste. But is there a reason this is a macro and not a function?
Those macros can't be debugged so it could be a little annoying in the future.

DMaroo replied:
We need to pass in different ops, of which each have their own `x86_il_##op##_with_rmode` function. I could probably have a macro which converts the op to the function name and passes that function pointer into a normal function, which can then call that function pointer. That would reduce the code in the macro, but would be (slightly) more convoluted. I would prefer it though.

Rot127 replied:
It's fine for me then.

DMaroo committed:
Add `RZ_IPI` annotation for all the functions exposed through the header

DMaroo committed:
Add `RZ_OWN`, `RZ_BORROW` and `RZ_NONNULL` annotations

DMaroo committed:
Add Doxygen doc for `ctx` argument

XVilka reviewed (changes requested):
Many newly added tests still fail:
```
[XX] db/asm/x86_64 <asm> faddp
-- <asm> faddp ---> dec1 ---> <IL>
-- assembly
-- IL
--- expected
+++ actual
@@ -1,1 +1,0 @@
-(seq (set _rmode (cast 2 false (>> (bv 8 0xa) (var cwd) false))) (set st1 (fbits (ite (== (var _rmode) (bv 2 0x0)) (+. rne (float 2 (var st0) ) (float 2 (var st1) )) (+. rtz (float 2 (var st0) ) (float 2 (var st1) ))))) (set swd (| (<< (cast 16 false (+ (cast 3 false (>> (var swd) (bv 8 0xb) false)) (bv 3 0x1))) (bv 8 0xb) false) (& (bv 16 0xc7ff) (var swd)))) (set st0 (fbits (float 2 (var st1) ))) (set st1 (fbits (float 2 (var st2) ))) (set st2 (fbits (float 2 (var st3) ))) (set st3 (fbits (float 2 (var st4) ))) (set st4 (fbits (float 2 (var st5) ))) (set st5 (fbits (float 2 (var st6) ))) (set st6 (fbits (float 2 (var st7) ))) (set swd (| (<< (ite (== (cast 3 false (>> (var swd) (bv 8 0xb) false)) (bv 3 0x0)) (bv 16 0x1) (bv 16 0x0)) (bv 8 0x9) false) (& (bv 16 0xfdff) (var swd)))))

WARNING: x86_il_set_st_reg_ctx: assertion 'val && ctx && x86_il_is_st_reg(reg)' failed (line 1325)
WARNING: x86_il_set_st_reg_ctx: assertion 'val && ctx && x86_il_is_st_reg(reg)' failed (line 1325)
WARNING: x86_il_set_st_reg_ctx: assertion 'val && ctx && x86_il_is_st_reg(reg)' failed (line 1325)
WARNING: x86_il_set_st_reg_ctx: assertion 'val && ctx && x86_il_is_st_reg(reg)' failed (line 1325)
WARNING: x86_il_set_st_reg_ctx: assertion 'val && ctx && x86_il_is_st_reg(reg)' failed (line 1325)
WARNING: x86_il_set_st_reg_ctx: assertion 'val && ctx && x86_il_is_st_reg(reg)' failed (line 1325)
WARNING: x86_il_set_st_reg_ctx: assertion 'val && ctx && x86_il_is_st_reg(reg)' failed (line 1325)
```

https://github.com/rizinorg/rizin/actions/runs/7448815863/job/20264020044?pr=3865#step:12:79

DMaroo commented:
I'll have a look at them. I'll be finishing up this PR by adding an emulateme test by the end of the week. Slightly busy right now.

DMaroo committed:
Add description for the `EXEC_WITH_RMODE` macro

DMaroo committed:
Move the non-null check for `ctx` inside the valid branch

    * When we pop the FPU stack we pass in a `NULL` context since no
      resizing would be needed, which causes a failure if the non-null
      check for the `ctx` is outside the branch
    * Fixes failing tests in `db/asm`

XVilka commented on librz/analysis/arch/x86/x86_il.c:
```diff
@@ -240,7 +258,75 @@ x86_il_ins x86_ins[X86_INS_ENDING] = {
+	[X86_INS_FNSTCW] = x86_il_fnstcw,
+	[X86_INS_FNSTSW] = x86_il_fnstsw,
+	[X86_INS_FNCLEX] = x86_il_fnclex,
+	[X86_INS_FLDENV] = x86_il_unimpl,
```
Maybe isolate unimplemented instructions somehow? E.g. put them at the end?

XVilka reviewed (changes requested):
Amazing job! 
Please fix those bindgen warnings as well:
```
Mismatched function annotation for x86_il_set_floating_operand_bits_ctx at <rizin/librz/analysis/arch/x86/common.h:190:22> : RZ_OWN (was None at <rizin/librz/analysis/arch/x86/common.c:1553:22>
Mismatched function annotation for x86_il_set_floating_operand_bits_ctx at <rizin/librz/analysis/arch/x86/common.h:190:22> : RZ_NONNULL (was None at <rizin/librz/analysis/arch/x86/common.c:1553:22>
Mismatched function annotation for x86_il_set_floating_operand_bits_ctx at <rizin/librz/analysis/arch/x86/common.h:190:22> : RZ_BORROW (was None at <rizin/librz/analysis/arch/x86/common.c:1553:22>
```

DMaroo committed:
Minor bug fixes + make the annotations less strict than they need to be

DMaroo commented:
Converted to draft because I gotta add the emulateme test for floating-point instructions. This PR should not be merged before that.

DMaroo committed:
Minor bug fixes

    * Incorrect implementation for `RET` and `CALL`
    * Swapped arguments for shift right

DMaroo committed:
Fix asm tests for `JMP` and `RET` instructions

XVilka commented:
@DMaroo a gentle ping on this one.

DMaroo committed:
Fix indexing error for `FXCH`

XVilka commented:
@DMaroo rebase please

XVilka commented on librz/analysis/arch/x86/common.c:
Remove this, we do not support Capstone 3 anymore, lowest one is CS4

XVilka reviewed (changes requested):
Even though there is no emulateme test, I propose merging this before the release, just mention that this support is experimental in the release notes, and will be improved in the next one. What do you think?

DMaroo committed:
Review changes

    * Move all unimplemented instructions to the end
    * Remove support for Capstone version less than 4

XVilka committed:
Merge branch 'dev' into x86-il-floating-point

DMaroo commented:
I agree about getting this merged. It is incorrect for some floating-point operations (due to the 80-bit floating-point shenanigans), but I'm confident that the rest of the instructions function as expected.
